service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/amqps/AmqpReceive.java

```diff
@@ -95,26 +75,14 @@ public synchronized FeedbackBatch receive(long timeoutMs) throws IOException, In
         feedbackBatch = null;
         if  (amqpReceiveHandler != null)
         {
-            // Codes_SRS_SERVICE_SDK_JAVA_AMQPRECEIVE_12_005: [The function shall initialize the Proton reactor object]
-            this.reactor = Proton.reactor(this);
-            // Codes_SRS_SERVICE_SDK_JAVA_AMQPRECEIVE_12_006: [The function shall start the Proton reactor object]
-            this.reactor.setTimeout(REACTOR_TIMEOUT);
-            this.reactor.start();
-            
+            log.info("Receiving on feedback receiver for up to {} milliseconds", timeoutMs);
+
             // Codes_SRS_SERVICE_SDK_JAVA_AMQPRECEIVE_12_007: [The function shall wait for specified timeout to check for any feedback message]
-            long startTime = System.currentTimeMillis();
-            long endTime = startTime + timeoutMs;
-            
-            while(this.reactor.process())
-            {
-                if (System.currentTimeMillis() > endTime) break;
-            }
-            
-            // Codes_SRS_SERVICE_SDK_JAVA_AMQPRECEIVE_12_008: [The function shall stop and free the Proton reactor object]
-            this.reactor.stop();
-            this.reactor.process();
-            this.reactor.free();   
-            this.amqpReceiveHandler.receiveComplete();
+            new ReactorRunner(this.amqpReceiveHandler, "AmqpFeedbackReceiver").run(timeoutMs);
+
+            log.trace("Feedback receiver reactor finished running, verifying that the connection opened correctly");
+            this.amqpReceiveHandler.verifyConnectionOpened();
```

Comment:
At this point, the reactor has stopped running and the reactor thread has been freed. Are the links/session/connection still open?

Reply:
In this implementation, yes, the links/session/connection were never closed. I've fixed this in the next iteration though. From the reactor runner, we need to signal to close the links first before we stop the reactor.